core/src/main/java/org/apache/iceberg/hadoop/HadoopInputFile.java

```diff
@@ -241,4 +272,118 @@ public void setNativeCryptoParameters(NativeFileCryptoParameters nativeCryptoPar
   public String toString() {
     return path.toString();
   }
+
+  protected void setContentCache(CachingHadoopFileIO.ContentCache cache) {
+    this.contentCache = cache;
```

Comment:
We can probably move the cache existence check here and additionally set `length` if appropriate.